### PR TITLE
Make CI work

### PR DESCRIPTION
**Pull request by haesleinhuepf**
I'm creating this PR as attempt to make github CI work. 

Please don't review yet! :-)

### CODE REVIEW

**Comment by haesleinhuepf** [comment]
Hi Stephane @StRigaud ,

I'm experiencing something suspicious here. Could you maybe have a look at the last two commits? in https://github.com/clEsperanto/pyclesperanto/pull/6/commits/89b9f3dad2b62fa23359e691f12c88c7321f1693 one test works that just uses `pyclesperanto_prototype`. After activating the build of `pyclesperanto` in https://github.com/clEsperanto/pyclesperanto/pull/6/commits/691c8044afb6d3cc663385218db7bc0b288fa005 , the same test fails. I'm having a hard time understanding, why so as these both libraries are independent. Do you have any idea?

Btw. pyclesperanto builds like a charm on my machine and the tests run. Very cool!

**Comment by StRigaud** [comment]
> Btw. pyclesperanto builds like a charm on my machine and the tests run. Very cool!

Great :smiley: 

> I'm experiencing something suspicious here. Could you maybe have a look at the last two commits? in [89b9f3d](https://github.com/clEsperanto/pyclesperanto/commit/89b9f3dad2b62fa23359e691f12c88c7321f1693) one test works that just uses `pyclesperanto_prototype`. After activating the build of `pyclesperanto` in [691c804](https://github.com/clEsperanto/pyclesperanto/commit/691c8044afb6d3cc663385218db7bc0b288fa005) , the same test fails. I'm having a hard time understanding, why so as these both libraries are independent. Do you have any idea?

I will have a look at it. Do you have any error returned by `pyclesperanto_prototype`? or from `pyclesperantor`?

My first guess would be a possible conflict on how pyopencl and CLIc connect and uses OpenCL. That is the only link between the two libraries.

**Comment by StRigaud** [comment]
So, I dont know how to solve but one issue here is that we do not use the same OpenCL install etc. and we are mixing stuff together.

We need to setup 3 things:

1. headers (cl.h, opencl.hpp, etc.)
2. icd-loader (libOpenCL.o.1, or something like that)
3. icd and a opencl device (something that actually can run opencl)

1 and 2 are need at compile time, 3 is needed at run time.

1 is shipt with CLIc.
2 is suposely taken care by `sudo apt fun`, but i dont think it works with pyopencl and pocl, hence `pycle_p` failing
3 pycle is failing because sudo apt fun does not provide any device and pocl which is, is not used.

pocl is installing a `libOpenCL.so.1` (an icd-loader) so we should be able to build both CLIc and pyopencl against it and run both on pocl.
All we need is to tell CMake to go fetch the opencl lib in the conda env

From there, both `pycle` and `pycle_p` will work against the same install of OpenCL and we can try debugging the stuff

**Comment by StRigaud** [comment]
So, I kind of stuck too, hence I did some local test

I've clean up my environment and restart from scratch.

### Test 1: install pocl, pyopencl, pyclesranto-prototype first, then pyclesranto:

First install using conda and conda-forge `pocl`, `pyopencl`, and `pyclesperanto-prototype`:
```python
from pyclesperanto_prototype import available_device_names
print(available_device_names())
```
returns `['pthread-Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz']`. It works and detect my CPU as an opencl device (thanks to pocl i guess) but no mention on by GPU.

Then compiling `pyclesperanto`:
```python
from pyclesperanto import gpu
print(gpu().info())
```
returns `[NVIDIA CUDA - NVIDIA GeForce GTX 1050 Ti] ...`.
then re-running `pycleseranto-prototype` fail `ImportError: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1: version `OPENCL_2.2' not found`.
removing `pyclesperanto` and `pyclesperanto-prototype` can run again.

result: pyclesperanto break pyopencl and pocl combo.


### Test 2: install pyclesranto first, then pocl, pyopencl, pyclesperantor-prototype:

First install `pyclesperanto`:
```python
from pyclesperanto import gpu
print(gpu().info())
```
returns `[NVIDIA CUDA - NVIDIA GeForce GTX 1050 Ti] ...`.

then install `pocl`, `pyopencl`, and `pyclesperanto-prototype`
```python
from pyclesperanto_prototype import available_device_names
print(available_device_names())
```
returns `['pthread-Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz']`

then testing `pyclesperanto' again:
```python
from pyclesperanto import gpu
print(gpu().info())
```
returns `[Portable Computing Language - pthread-Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz]`

result: all is pointing to the CPU device enabled by pocl. The GPU disappear though ...

### Test 3: install pyclesperanto and pyclesperanto-prototype without pocl

in short, pyclesperanto works but pyopencl cannot find device (`pyopencl._cl.LogicError: clGetPlatformIDs failed: PLATFORM_NOT_FOUND_KHR`) with the message:
```
WARNING: No ICDs were found. Either,
- Install a conda package providing a OpenCL implementation (pocl, oclgrind, intel-compute-runtime, beignet) or 
- Make your system-wide implementation visible by installing ocl-icd-system conda package.
```

running `conda install ocl-icd-system`, pyclesperanto is all good (still pointing to my GPU), pyclesperanto-prototype is failing with `ImportError: /usr/lib/x86_64-linux-gnu/libOpenCL.so.1: version `OPENCL_2.2' not found`

## tl:dl

Still fuzy and unclear but it is linked to which ICD are available and visible. The install order matters here though it should not.

pyclesperanto seems to see both GPU and CPU depending on the order of install but not both at the same time (this is annoying).
pyclesperanto-prototype only see CPU and fail to see GPU (why? what do I miss?). 

I am getting lost in all thos ICD, ICD-loader, etc ...

**Comment by StRigaud** [comment]
@haesleinhuepf I am kind of stuck here too. Locally I manage to make some stuff works fine but nothing to solve all this in the CI. I am running out of ideas and getting lost in the ICD and loaders and versions.

**Comment by StRigaud** [comment]
now `pyclesperanto` CMake search if it find a `libOpenCL.so` file in the python environment
```CMake
file(GLOB OpenCL_LIBRARY ${Python_LIBRARY_DIRS}/libOpenCL.so) 
if(OpenCL_LIBRARY STREQUAL "")
    message(STATUS "No ICD found in virtual environment.")
else()
    set(OpenCL_LIBRARIES ${OpenCL_LIBRARY})
    find_package(OpenCL 1.2 PATHS ${Python_LIBRARY_DIRS} ${Python_INCLUDE_DIRS})
    set(OpenCL_FOUND TRUE)
endif()
```
If it is found, then CLIc will be compiled against it, otherwise it will compile against the one installed in the system if found (aka in PATH).

Now, `libOpenCL.so` is an ICD-loader which is provided by the package `ocl-icd`. With this `pyclesperanto` should build but test will fail as no ICD will be found at runtime.

Now, from the conda env, we have two options to provide ICDs: make the system ICDs (hardware GPU) visible using the package `ocl-icd-system` or install in the environment an ICD like `pocl` which provide `pthread` for the CPU.
- with `ocl-icd-system`, it detect my GPU, run in a python shell, and it run with pytest.
- my cmake compilation detect pthread but my scikit-build (using `pip install -e .`) compilation detect my GPU instead of pthread.

I believe the last point is the issue for making the CI work.

For CMake to know that it is inside a conda environment, the configuration must be called inside the env by command line. I fear the scikit-build prevent this detection. but I cannot see the compilation verbose of scikit-build to confirm it.

More tomorrow after a day of rest :exploding_head:

**Comment by StRigaud** [comment]
Ok, I manage to reproduce the error `ImportError: libOpenCL.so.1: cannot open shared object file: No such file or directory` locally.

I remove all opencl installation on my OS (GPU still here, but I removed all package linked to ICD and headers).
I only install in my conda env the package `pocl` which is shipt with he package `ocl-icd` (that install a `libOpenCL.so` in the conda env).

At compilation type, `pyclesperanto` find the `libOpenCL.so` install by `ocl-icd` and link to it. All is good.
At execution time it run the error and fail.

If I do the same but with `pyopencl` and `pyclesperanto-prototype`, their test pass and the platform `['pthread-Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz']` is found.

I do not understand why `pyopencl/pyclesperanto-prototype` see pthread-Intel and not `pyclesperanto`:

- option 1: I am missing something during CLIc compilation
- option 2: I am missing a link with pocl
- option 3: I am missing a package that is provided by pyopencl to make the link with pocl.

Basically I am missing a something to reach out to pocl

**Comment by StRigaud** [comment]
Strange stuff:

If running test for pyclesperanto-prototype first, then pyclesperanto, both manage to find the platform (pthread-Intel) (why, how, .... :man_shrugging:)

However the actual execution test still fail because returning wrong output, similar to when running it with oclgrind.

**Comment by StRigaud** [comment]
I found how to solve the issue of pocl and oclgrind returning wrong values.

When creating a buffer (same for image) we can use various flag for memory management.
```c++
cl_mem_flags mem_flags = CL_MEM_READ_WRITE  | CL_MEM_USE_HOST_PTR;
t_buffer = cl::Buffer (this->Context(), mem_flags, t_bitsize, t_ptr, &error);
```

`CL_MEM_USE_HOST_PTR` appears to not be compatible with the python wrapper and CPU run. Removing it solve the isse of wrong value returned raised in
- #1 

However the error `ImportError: libOpenCL.so.1 still an issue` remains ...

**Comment by StRigaud** [comment]
closing the PR as this is maybe not a thing anymore and we may go towards different approach
